csrc/multidevice/executor.cpp

```diff
@@ -49,46 +51,95 @@ std::unique_ptr<Fusion> copyFusionAndChangeOutputs(
+  // Infer a topologically ordered traversal of the segmented fusion to
+  // determine the order for launching the kernels/comms
+  RuntimeWorkSpace workspace;
+  prepareRuntimeOrder(staged_fusion.get(), workspace);
+
+  // Create the HostIrContainer representing the host program. Each segment of
+  // the segmented fusion will be translated to a HostIR
+  auto hic = std::make_unique<hir::HostIrContainer>();
+  FusionGuard fg(hic.get());
+  IrCloner ir_cloner(hic.get());
+  auto clone =
+      [&ir_cloner](const std::vector<Val*>& vals) -> std::vector<Val*> {
+    std::vector<Val*> cloned_vals(vals.size());
+    std::transform(
+        vals.begin(), vals.end(), cloned_vals.begin(), [&ir_cloner](Val* val) {
+          return ir_cloner.clone(val);
+        });
+    return cloned_vals;
+  };
 
-  staged_fusion_ =
-      SegmentCandidateFinder::segment(std::move(fusion), nullptr, options);
-
-  for (auto group : staged_fusion_->groups()) {
+  for (auto group : workspace.group_run_order) {
+    std::vector<Expr*> host_exprs;
     NVF_ERROR(!group->exprs().empty(), "invalid segmentation");
-    is_resharding_[group] = std::any_of(
+    bool is_resharding = std::any_of(
         group->exprs().begin(), group->exprs().end(), [](auto expr) {
           return isResharding(expr);
         });
-    NVF_ERROR(
-        !is_resharding_[group] || group->exprs().size() == 1,
-        "Communications cannot be fused");
-    auto expr = group->exprs().at(0);
-    should_run_[group] = involvedDevices(expr).count(comm_.deviceId());
+    if (involvedDevices(group->exprs().at(0)).count(comm_.deviceId()) == 0) {
+      continue;
+    }
+    if (!is_resharding) {
+      auto host_unit = IrBuilder::create<hir::HostUnit>(
+          staged_fusion->makeFusion(group).second);
+      auto post_on_stream = IrBuilder::create<hir::PostOnStream>(
+          host_unit, clone(group->inputs()), clone(group->outputs()));
+      hic->pushBackTopLevelExprs(post_on_stream);
+    } else {
+      NVF_ERROR(
+          group->exprs().size() == 1,
+          "Communication segments must contain only one Expr");
+      std::vector<Communication*> communications =
+          lowerCommunication(ir_cloner.clone(group->exprs().at(0)));
+      for (Communication* communication : communications) {
+        auto wait = IrBuilder::create<hir::Wait>(communication);
+        hic->pushBackTopLevelExprs(communication);
+        hic->pushBackTopLevelExprs(wait);
+      }
+    }
+  }
+  for (auto input : staged_fusion->inputs()) {
+    hic->addInput(ir_cloner.clone(input));
+  }
+  for (auto output : staged_fusion->outputs()) {
+    hic->addOutput(ir_cloner.clone(output));
   }
-  // prepare the order in which to launch the kernels/comms
-  prepareRuntimeOrder(staged_fusion_.get(), workspace_);
+
+  // Create the HostIrExecutor representing the host program
+  host_ir_executor_ =
+      std::make_unique<hir::HostIrExecutor>(std::move(hic), &comm, params);
```

Comment:
```suggestion
      std::make_unique<hir::HostIrExecutor>(std::move(hic), &comm, std::move(params));
```
to avoid an unnecessary copy.

Reply:
That's what I did originally but the linter CI complained
https://github.com/NVIDIA/Fuser/actions/runs/9683775298/job/26720018890

Reply:
That complain is actually correct. There has been several places where `const T&` could have been used to save copies, which I'd probably do. I thought `T` plus `std::move` can save copies as well but the clangtidy reminded me that `HostIrExecutorParams` is a trivially copyable type that does not have a move constructor. So `std::move` is as expensive as a copy.